Add user events for useful events for optimizing statusline performance

mehalter opened the pull request:
When adding status to the statusline many statusline providers support caching and updating on events. This adds some helpful user events which can be used to optimize the use of `get_status` so that it's not calling `get_status` on every clock tick. I also went ahead and added a useful section to the README to help expose this feature to the user. This also provides a means for the user to create mappings only when the codeium server is connected and then removing them when the codeium server disconnects.


Thanks so much for a great plugin it works marvelously!

EDIT: This also adds a `toggle_buffer` command as that seems pretty useful as well 😄 

Closes #21

luxus commented:
it's a great idea and i'm super into it

monkoose commented:
Hi, again. It was bed time for me 💤, I will review it shortly.

monkoose commented on lua/neocodeium/commands.lua:
```diff
@@ -181,10 +185,12 @@ function M.restart()
-   if vim.g.neocodeium_enabled and not server.pid then
-      server:run()
-   end
+   M[vim.g.neocodeium_enabled and "disable" or "enable"]()
```
This is hard no from me. 

1. I don't like "smart" code. I don't care if it would span few more lines, but this is unreadable, doesn't allow to jump to definition of those functions without manually searching.
2. It doesn't act like the old code, yes in the current plugin state it doesn't matter because I have added that g.neocodeium_enabled is always present and without user manually deletes it all would be fine, but still on missing variable this code behaves wrong. 

But at the same time you are right that this is a code duplication, so it should be addressed. 

I like to name variables with the on/off meaning as non-negative (bad English not sure how to explain this right), so I have named them enabled in the plugin. But maybe because they are not documented and you can say hidden from the user, it is better name them with negative meaning as vim.g and vim.b `neocodeium_disabled`? What is your opinion. 


TL;DR this, and especially buffer toggle should be rewritten using if statement.

mehalter replied:
Awesome, thanks for the feedback! I revised the commits to use the check against false and moved them to `if`/`else` statements

monkoose commented:
My opinion that all such doautocmds except ServerConnected CompletionDisplayed and CompletionCleared should be renamed as one group StatusChanged or something, unless you proved the reason why they shouldn't. And after ServerConnected should be this StatusChanged doautocmd also. So users can watch status change with only one autocmd in their configs.

As of your comment on reddit (if I'm guess right and it was you), neovim still process the code and writes UI on the same core if I'm not mistaken. And any of "scheduled", "timed", "coroutined", "asynced" function execution still bogs it in some way. Or I'm wrong?

monkoose commented:
I think that this line should also have StatusGhanged doautocmd, to notify when server is in connecting status
https://github.com/monkoose/neocodeium/blob/6d2df2dbedce6f53a9b09cfd5ff549d834269202/lua/neocodeium/server.lua#L108

monkoose commented:
Ok. Thats all.

Anyway maybe I'm stupid, but I still don't understand how you want to utilize CompletionDisplayed and CompletionCleared autocmds for your opened issue. Or they don't allow what you want?

mehalter commented:
I can think of a lot of reasons to keeping these events separate. In my personal configuration where I am testing using this code I use them for many different tasks. `ServerConnected` is used to tell my editor to create bindings for NeoCodeium, and I do different things on `CompletionDisplayed` and `CompletionCleared`. The nice thing about patterns for the user events is you can always do wildcards in the pattern matching to easily get all of the statuses at one time that you care about. So the user can make an autocommand with the pattern `NeoCodeium*` or `NeoCodeiumServerConnected|NeoCodeiumCompletionDisplayed` or whatever they want.

Also in regards to the issue I opened previously (#21), this does achieve what I want there with only `CompletionDisplayed` (simply if a completion is displayed I have an autocommand that closes my `nvim-cmp` completion).

I did some testing and you are correct as I started doing more testing the `scheduling doesn't make things run asynchronously and just goes in a different place in the execution. If the user wants they could do their own asynchronous execution using whatever methods they want from within the autocommands they create. The nice thing is the plugin doesn't create auto commands so there is no loss in performance out of the box and it would be up to the use to make their code performant. In most cases I imagine the operations people are doing when these events trigger will be very low cost. Stuff like clearing statusline cache for an element, creating/deleting mappings, etc.

Thanks so much for the review and I'm excited to contribute to this awesome project! I added the `ServerConnecting` event, this is actually much needed for correctly updating statusline components in more cases. I also reverted the "smart" code but still handled removing the duplication of the enable/disable functionality. I also removed the scheduling as this doesn't change anything other than make the events execute in potentially the wrong order/wrong time. Should be good to go 😄

monkoose commented:
Sorry, maybe language issue or you get it wrong. I totally understand importance of ServerConnected, CompletionDisplayed and CompletionCleared. But why do others need to be different? How do you use them for statusline then? With big list that lists them all? Can you name any reason why they do need to have different name?

mehalter commented:
> Could you also address pending review. Or write your opinion why it is important.

Ah maybe I missed something, what is the pending review? I thought I addressed all of the comments and concerns. Sorry!

monkoose commented:
> Outside of my statusline cache I also use NeoCodeiumServerStopped to remove mappings

How do you stop it? With `require('neocodeium.server'):stop()`?

mehalter commented:
Yeah

Also what do you mean you updated the pending line?

monkoose commented:
You don't see it?
`   if vim.F.if_nil(vim.b.neocodeium_enabled, vim.g.neocodeium_enabled) == false then` this line

Is there any reason to complecate it that much?
- If vim.g is enabled, and we first introduce this vim.b there is not point of it it would always equal false
- If vim.g. is disabled there is no point of setting vim.b to true, because global state > local state (which just leads to confusion)

monkoose commented:
I also think that when it doesn't matter for all other events.
But `CompletionCleared` and `CompletionDisplayed` should be scheduled ones. I don't mind another utils function like `scheduled_event`. Aren't they stacked anyway so, they can't be invoked in the wrong order? 
Heavy login on this events could add input delay.

mehalter commented:
This is a pretty standard method for retrieving the current status of something buffer local vs global where the buffer local state is unknown. This function returns the first argument that is not nil. So in the case of the global state is configured and the buffer local state is not configured yet and the global state is being used, it guarantees that the new buffer local state is going to be the opposite of the current global state.

Say the user has not done any buffer toggling but toggled the global state to false, when they do a buffer toggle they want to toggle the buffer to be the opposite of it's current state which is disabled. So then the new buffer state should be enabled. This makes sure it's that case.

Does that make sense?

Also cool I will add scheduling to the displayed and cleared events. Thanks!

mehalter commented:
Oh wait, I see what you are saying, you are saying the buffer local setting does not take precedence over the global setting? Is that intended? I suppose I didn't realize that since that's the opposite of how buffer local/global options work in vim/neovim in general. If that's the case, then maybe I just remove the buffer toggling all together in this PR and it can be added separately. Sorry for the noise there!

monkoose commented:
> If that's the case, then maybe I just remove the buffer toggling all together in this PR and it can be added separately

I don't mind really, not following super rules about PR do one thing.
But maybe it is really better to split `toggling` to different PR, it up to you. When you address all my questions I will quick test it and merge anyway.

monkoose commented:
Yes it is intended that global disable fully disables the plugin.

mehalter commented:
Ok, I went ahead and added scheduling for the renderer related events.

I also removed the buffer toggling for now as I am not fully versed in the handling of the global vs local state of buffers so I think I'm probably not a good candidate to add that feature at the moment. If it's still desired I could take some time to read through and digest the precedence of stuff and take another stab at it in a follow up PR. Sorry about that!

monkoose commented:
Ok, thanks. I will do it myself later.

mehalter commented on lua/neocodeium/commands.lua:
```diff
@@ -181,10 +185,12 @@ function M.restart()
+end
+
+function M.toggle_buffer()
+   -- if a buffer local control is set do the opposite of it, if not do the opposite of the current global state
```
Ohhhh I just got these, I think you forgot to submit the review so these comments were pending unable to be seen 😅 sorry!

mehalter commented on lua/neocodeium/commands.lua:
```diff
@@ -181,9 +185,19 @@ function M.restart()
+
+function M.toggle_buffer()
+   -- if a buffer local control is set do the opposite of it, if not do the opposite of the current global state
+   if vim.F.if_nil(vim.b.neocodeium_enabled, vim.g.neocodeium_enabled) == false then
```
Ah this comment as well, I see now what I was missing

monkoose replied:
Nah, that I'm who was doing it wrong.

Thanks again.
